### DISCUSSION TITLE
Having troubles with connection

### DISCUSSION THREAD

**Discussion by expecticament** [Pryaxis/TShock#1717] [General]
My friends can't connect to my TShock server, they just stuck on "Connecting to...". We tried to use all of my IPs, but nothing works. (port forwarding is turned on, port 7777)

**Comment by QuiCM** [comment]
Are connections to/from Terraria server allowed through your firewall?

**Reply by expecticament** [reply]
My firewall is turned off

**Comment by hakusaro** [comment]
Question: You've been using an ip that you know is on the public internet, right? Like from [ip4.me](http://ip4.me/)? For port forwarding to work, you also need the gateway/router configured for TCP on 7777 to your local IP (from the command line: `ipconfig /all`, and look for your local IP). All of these have to be IPv4 IPs, like `1.2.3.4`. You know you're looking at a local IP if it starts with `192.168.x.x`, `10.0.x.x`, `172.16.x.x.`.

**Reply by expecticament** [reply]
My public IP is 128.71.x.x. We are already tried my local IP. But it doesn't works too(stuck on connecting to...) It's interesting that official Terraria server from game folder works perfectly and everyone can connect with my public IP. (I'm not good enough at English, sorry for my grammar mistakes)

**Reply by bartico6** [reply]
You shouldn't post your public IP on the internet (publicly at least).
If your public IP doesn't work with tShock, but with the official dedicated server it does, then it's a matter of port forwarding. The stock Terraria server uses UPnP punch-through to get the port open automatically, such functionality is **not** present in tShock as far as I'm aware.

**Reply by bartico6** [reply]
If your local IP (local as in intranet, inside your home network) doesn't work, then it might also be an issue with your computer's firewall settings. You should try connecting to 127.0.0.1 and see if that works. Generally speaking though, if there's a Windows Firewall prompt like the one below, you should allow it.
![Example firewall prompt](https://i.quake.tf/1588315155-64582146.png)

**Reply by hakusaro** [reply]
I believe in @bartico6's last post, he meant "try connecting to 127.0.0.1"?

**Reply by bartico6** [reply]
That is correct, I'll amend the post to reflect that

**Reply by expecticament** [reply]
There's no any problems if I connect to my own server with "localhost", "127.0.0.1".

**Reply by moisterrific** [reply]
Then it’s most likely not a TShock issue.

**Comment by hakusaro** [comment]
I removed your IP unless you really want it hanging around. You're definitely using UPnP. The trick is that your local IP will change a lot on your internal network unless you have a static IP assignment internally. Port forwarding works by taking incoming packets from the Internet and forwarding them to a known destination. If you don't have a known destination, like if your computer lost its IP, then you won't have the same IP anymore.

In other words:

1. Your friends must connect to your public IP.
2. Your router/gateway must forward port `7777` to your computer's port `7777` on your _local IP_.

To get a static local IP, you can edit your network settings to assign one, or you can use a "DHCP reservation" on the router/gateway. If you turn on a DHCP reservation, it'll always give your computer the same IP. That way you can turn it on, find your local IP, and then map the port there. You can follow these instructions to both set a static IP and to port forward based on your router: https://portforward.com/help/terraria/.

**Reply by expecticament** [reply]
OK, thanks a lot. I will try this one.

**Reply by hakusaro** [reply]
Any luck @ShadowNotFake?

**Reply by expecticament** [reply]
I have not tried it yet,  because I am not at home. I will write in 2-4 days, when I will back.

**Reply by expecticament** [reply]
It works! Thanks a lot for your help:D
